server/src/main/java/org/elasticsearch/common/settings/Settings.java

```diff
@@ -158,7 +159,10 @@ private Object convertMapsToArrays(Map<String, Object> map) {
         for (Map.Entry<String, Object> entry : map.entrySet()) {
             if (isArray) {
                 try {
-                    int index = Integer.parseInt(entry.getKey());
+                    final String key = entry.getKey();
+                    // check whether string may be an integer first (mostly its not) to avoid the slowness of parseInt throwing in a hot
+                    // loop
+                    int index = Numbers.isPositiveNumeric(key) ? Integer.parseInt(key) : -1;
```

Comment:
I really like this fix, exception throwing/catching is very expensive. I would suggest one change only. If the key for some reason was null (I'm not sure it's possible for ES settings, but it's possible for j.u.Map) then the old code would catch the NumberFormatException and set isArray to false. The modified code would hit NullPointerException in Numbers.isPositiveNumeric and fail to execute the remainder of the code.

Reply:
I think we're safe from a `null` here. If we had a map with `null` keys it would fail serialization over the wire and a number of other spots outright (+ it would just be a bug to begin with :)).

Reply:
That makes sense, better to let it fail then with NullPointerException!